### PR TITLE
Enforce the block size limit

### LINKED ISSUE #2133: Add support block size limit
Currently, the block is limited by the gas, and with the `63` gas per byte, the maximum size of the block can be `30000000/63 = 476_190` byte.

But since we want to have a hard limit 128kb or 256kb, we either need to increase the gas per byte, or add a new block size limit variable.

Support the block size limit requires:
- Add a new `block_size_limit` field to the consensus parameters.
- Updating the executor's logic to get transactions based on the remaining size.
- Update the transaction pool to support the limitation by size.

### PR DESCRIPTION
**Pull request by rafal-ch**
Closes #2133

## Linked Issues/PRs
Built on top of: https://github.com/FuelLabs/fuel-core/pull/2188

## Description
This PR adds the handling and enforcement of then new consensus parameter added [here](https://github.com/FuelLabs/fuel-core/pull/2188).

Changes summary:
- `TransactionsSource` and `transaction_selector` use the new `block_transaction_size_limit` parameter.
- `ExecutionData` contains `used_size` in addition to `used_gas`
- Introduces integration tests covering the expected behavior
- E2E tests affected by the new limit are updated

## Checklist
- [X] Breaking changes are clearly marked as such in the PR description and changelog
- [X] New behavior is reflected in tests

### Before requesting review
- [X] I have reviewed the code myself

### CODE REVIEW

File: crates/services/upgradable-executor/wasm-executor/src/ext.rs [hunk 1]
```diff
@@ -148,8 +151,13 @@ pub fn input(size: usize) -> anyhow::Result<InputDeserializationType> {
...
 /// Gets the next transactions by using the host function.
-pub fn next_transactions(gas_limit: u64) -> anyhow::Result<Vec<MaybeCheckedTransaction>> {
-    let next_size = unsafe { host::peek_next_txs_size(gas_limit) };
+pub fn next_transactions(
+    gas_limit: u64,
+    block_transaction_size_limit: u64,
+) -> anyhow::Result<Vec<MaybeCheckedTransaction>> {
+    // TODO[RC]: Peek, check size. If fits, continue. If too large, peek next (?)
+    let next_size =
+        unsafe { host::peek_next_txs_size(gas_limit, block_transaction_size_limit) };
```
Comment (rafal-ch) [inline crates/services/upgradable-executor/wasm-executor/src/ext.rs:160]: Note to self:
We should most likely keep the `peek_next_txs_size` and introduce new function for the limits.

Comment (rafal-ch) [reply crates/services/upgradable-executor/wasm-executor/src/ext.rs]: This has been taken care of by merging changes introduced here: https://github.com/FuelLabs/fuel-core/pull/2199

File: crates/services/txpool/src/transaction_selector.rs [hunk 2]
```diff
@@ -6,27 +6,43 @@ use fuel_core_types::{
...
-    // Pick as many transactions as we can fit into the block (greedy)
+    // Pick as many transactions as we can fit into the block (greedy),
+    // respecting both gas and size limit
     includable_txs
-        .filter(|tx| {
-            let tx_block_space = tx.max_gas();
-            if let Some(new_used_space) = used_block_space.checked_add(tx_block_space) {
-                if new_used_space <= max_gas {
-                    used_block_space = new_used_space;
+        .filter(move |tx| {
```
Comment (xgreenx) [inline crates/services/txpool/src/transaction_selector.rs:28]: It is not part of this PR, but filtering *each* entry of the transaction pool is too expensive. We should rework it and stop iteration when we filled our limits.

File: crates/services/txpool/src/transaction_selector.rs [hunk 3]
```diff
@@ -6,27 +6,43 @@ use fuel_core_types::{
...
+            let tx_block_gas_space = tx.max_gas();
+            let tx_block_size_space = tx.metered_bytes_size() as u64;
+
+            let maybe_new_used_block_gas_space =
+                used_block_gas_space.checked_add(tx_block_gas_space);
+            let maybe_new_used_block_size_space =
+                used_block_size_space.checked_add(tx_block_size_space);
```
Comment (xgreenx) [inline crates/services/txpool/src/transaction_selector.rs:35]: I think we can use `saturating_add` isntead

File: tests/tests/tx.rs [hunk 4]
```diff
@@ -190,6 +198,142 @@ async fn dry_run_above_block_gas_limit() {
...
+        let srv = FuelService::new_node(config).await.unwrap();
+        let client = FuelClient::from(srv.bound_address);
+
+        // Submit all transactions
+        for (tx, _) in &transactions {
+            let status = client.submit(tx).await;
+            assert!(status.is_ok())
+        }
+
+        let _ = tokio::time::sleep(Duration::from_millis(10)).await;
```
Comment (xgreenx) [inline tests/tests/tx.rs:273]: Why do we need to wait? You can use the `srv.shared_state.txpool.insert` function if you want to be sure that the transaction is inserted. But  `client.submit(tx).await` should work as well

File: tests/tests/tx.rs [hunk 5]
```diff
@@ -190,6 +198,142 @@ async fn dry_run_above_block_gas_limit() {
...
+    let srv = FuelService::new_node(config).await.unwrap();
+    let client = FuelClient::from(srv.bound_address);
+
+    // Submit all transactions
+    for (tx, _) in &transactions {
+        let status = client.submit(tx).await;
+        assert!(status.is_ok())
+    }
+
+    let _ = tokio::time::sleep(Duration::from_millis(10)).await;
```
Comment (xgreenx) [inline tests/tests/tx.rs:324]: The same here, do we need to wait if you use manual block production?

File: Cargo.lock [hunk 6]
```diff
@@ -1252,9 +1252,9 @@ checksum = "1fd0f2584146f6f2ef48085050886acf353beff7305ebd1ae69500e27c67f64b"
 
 [[package]]
 name = "bytes"
-version = "1.7.1"
+version = "1.7.2"
```
[the remark this answers is not in the document]
Comment (rafal-ch) [reply Cargo.lock]: Changes to `Cargo.lock` reverted in https://github.com/FuelLabs/fuel-core/pull/2195/commits/9bdea6d47bfac2cd9f1c3ba431134bb5f9ee2cab

File: crates/services/txpool/src/transaction_selector.rs (hunk 3 above)
Comment (rafal-ch) [reply crates/services/txpool/src/transaction_selector.rs]: You're right. Update in https://github.com/FuelLabs/fuel-core/pull/2195/commits/36ab508eaa109bc47eba6271ae1ba01fe3f5096a

File: tests/tests/tx.rs (hunk 4 above)
Comment (rafal-ch) [reply tests/tests/tx.rs]: It's a leftover from my initial test attempts. Removed in https://github.com/FuelLabs/fuel-core/pull/2195/commits/15a2cf7eaa1102cdfe93ed9eea8988a5b4278230.

cc: @MitchTurner - you may want to have a look at this.

File: tests/tests/tx.rs (hunk 5 above)
Comment (rafal-ch) [reply tests/tests/tx.rs]: See: https://github.com/FuelLabs/fuel-core/pull/2195#discussion_r1769589697

File: crates/services/txpool/src/transaction_selector.rs (hunk 2 above)
Comment (rafal-ch) [reply crates/services/txpool/src/transaction_selector.rs:28]: I created a follow-up issue https://github.com/FuelLabs/fuel-core/issues/2235

**Review by xgreenx — APPROVED** [review]

**Review by acerone85 — DISMISSED** [review]
Had a look at the code and the tests. 

It looks like the implementation for `OnceTransactionsSource` is still ignoring the `block_txs_size` parameter (https://github.com/FuelLabs/fuel-core/blob/79ca0d04d863433c30e208241f363790e8b8b63d/crates/services/executor/src/executor.rs#L192), but this could be done as a follow up. 

LGTM.

**Comment by rafal-ch** [comment]
> It looks like the implementation for `OnceTransactionsSource` is still ignoring the `block_txs_size` parameter

Summoning @xgreenx for a comment.

According to my current understanding, we don't necessarily need the support for `block_txs_size` in the `OnceTransactionsSource`.

**Comment by xgreenx** [comment]
> According to my current understanding, we don't necessarily need the support for block_txs_size in the OnceTransactionsSource.

Yes, `OnceTransactionsSource` shouldn't event exist in production, it is just a hack for dry run=)

File: crates/services/executor/src/executor.rs [hunk 7]
```diff
@@ -194,7 +194,7 @@ impl TransactionsSource for OnceTransactionsSource {
         &self,
         _: u64,
         transactions_limit: u16,
-        _: u32,
```
Comment (xgreenx) [inline crates/services/executor/src/executor.rs:197]: I missed that, but the block size should be `u32`.

File: crates/services/upgradable-executor/wasm-executor/src/ext.rs [hunk 8]
```diff
@@ -84,7 +84,7 @@ mod host {
         pub(crate) fn peek_next_txs_size(
             gas_limit: u64,
             tx_count_limit: u32,
-            size_limit: u32,
+            block_transaction_size_limit: u64,
```
Comment (xgreenx) [inline crates/services/upgradable-executor/wasm-executor/src/ext.rs:87]: It is a breaking change

**Review by xgreenx — CHANGES_REQUESTED** [review]

File: crates/services/executor/src/executor.rs (hunk 7 above)
Comment (rafal-ch) [reply crates/services/executor/src/executor.rs]: Updated in https://github.com/FuelLabs/fuel-core/pull/2195/commits/e748837be51c4d738912768c0aa33d605444118f

File: crates/services/upgradable-executor/wasm-executor/src/ext.rs (hunk 8 above)
Comment (rafal-ch) [reply crates/services/upgradable-executor/wasm-executor/src/ext.rs]: Updated in https://github.com/FuelLabs/fuel-core/pull/2195/commits/e748837be51c4d738912768c0aa33d605444118f

**Review by xgreenx — APPROVED** [review]

**Review by acerone85 — APPROVED** [review]
